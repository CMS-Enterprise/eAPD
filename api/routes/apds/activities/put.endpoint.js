const {
  getDB,
  getFullPath,
  login,
  request,
  unauthenticatedTest,
  unauthorizedTest
} = require('../../../utils.endpoint');

describe('APD activities endpoint | PUT /activities/:id', async () => {
  const db = getDB();
  beforeAll(() => db.seed.run());
  afterAll(() => db.destroy());

  const url = getFullPath('/activities');

  unauthenticatedTest('put', `${url}/1`);
  unauthorizedTest('put', `${url}/1`);

  describe('when authenticated as a user with permission', async () => {
    let cookies;
    beforeAll(async () => {
      cookies = await login();
    });

    it('with a non-existant activity ID', async () => {
      const { response, body } = await request.put(`${url}/9000`, {
        jar: cookies,
        json: true
      });

      expect(response.statusCode).toEqual(404);
      expect(body).toMatchSnapshot();
    });

    it(`with an activity on an APD in a state other than the user's state`, async () => {
      const { response, body } = await request.put(`${url}/4110`, {
        jar: cookies,
        json: true
      });

      expect(response.statusCode).toEqual(404);
      expect(body).toMatchSnapshot();
    });

    it('with an activity name that already exists for the APD', async () => {
      const { response, body } = await request.put(`${url}/4100`, {
        jar: cookies,
        json: {
          name: 'My Second Activity'
        }
      });

      expect(response.statusCode).toEqual(400);
      expect(body).toMatchSnapshot();
    });

    it('with an empty activity name', async () => {
      const { response, body } = await request.put(`${url}/4100`, {
        jar: cookies,
        json: {
          name: ''
        }
      });

      expect(response.statusCode).toEqual(400);
      expect(body).toMatchSnapshot();
    });

    it('with a valid activity', async () => {
      const { response, body } = await request.put(`${url}/4100`, {
        jar: cookies,
        json: {
          name: 'updated name',
          description: 'updated description'
        }
      });

      expect(response.statusCode).toEqual(200);
      expect(body).toMatchSnapshot();
    });
  });
});
