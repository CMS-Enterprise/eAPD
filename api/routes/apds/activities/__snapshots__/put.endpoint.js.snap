// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APD activities endpoint | PUT /activities/:id when authenticated as a user with permission with a non-existant activity ID 1`] = `undefined`;

exports[`APD activities endpoint | PUT /activities/:id when authenticated as a user with permission with a valid activity 1`] = `
Object {
  "alternatives": null,
  "contractorResources": Array [
    Object {
      "description": null,
      "end": null,
      "files": Array [
        Object {
          "id": 5002,
          "size": null,
        },
      ],
      "hourlyData": Array [],
      "id": 4300,
      "name": null,
      "start": null,
      "useHourly": false,
      "years": Array [],
    },
  ],
  "costAllocation": Array [],
  "costAllocationNarrative": Object {
    "methodology": null,
    "otherSources": null,
  },
  "description": "updated description",
  "expenses": Array [],
  "files": Array [
    Object {
      "id": 5000,
      "in": "here",
      "size": null,
      "some": "metadata",
    },
    Object {
      "id": 5004,
      "size": null,
    },
  ],
  "fundingSource": null,
  "goals": Array [
    Object {
      "description": "Be a super successful artist",
      "id": 4200,
      "objective": "Paint a pretty picture",
    },
    Object {
      "description": "Win a Nobel prize for physics",
      "id": 4201,
      "objective": "Discover a new particle",
    },
    Object {
      "description": "Go on Ellen",
      "id": 4202,
      "objective": "Learn to dance",
    },
  ],
  "id": 4100,
  "name": "updated name",
  "plannedEndDate": "1963-06-16",
  "plannedStartDate": "1970-04-11",
  "quarterlyFFP": Array [],
  "schedule": Array [],
  "standardsAndConditions": null,
  "statePersonnel": Array [],
  "summary": null,
}
`;

exports[`APD activities endpoint | PUT /activities/:id when authenticated as a user with permission with an activity name that already exists for the APD 1`] = `
Object {
  "action": "update-activity",
  "error": "activity-name-exists",
}
`;

exports[`APD activities endpoint | PUT /activities/:id when authenticated as a user with permission with an activity on an APD in a state other than the user's state 1`] = `undefined`;

exports[`APD activities endpoint | PUT /activities/:id when authenticated as a user with permission with an empty activity name 1`] = `
Object {
  "action": "update-activity",
  "error": "activity-name-invalid",
}
`;
